SFM.py

```diff
@@ -259,6 +260,101 @@ def output(self):
+
+        for i, frame_kps in enumerate(kps_per_frame):
+            self.db.add_keypoints(i+1, frame_kps)
+
+        for i, frame_desc in enumerate(desc_per_frame):
+            self.db.add_descriptors(i+1, frame_desc)
+
+        self.db.commit()
+
+        # adding matches and two view geo to db
+        kps_2d = [np.array([[y.pt[0], y.pt[1]] for y in x[0]], dtype=np.float64) for x in self.desc]
+
+        for i in self.good.keys():
+            if isinstance(i, tuple):
+                m = np.array([[x.queryIdx,x.trainIdx] for x in self.good[i]], dtype=np.uint32)
+                if len(m) > 0:
+                    self.db.add_matches(i[0]+1, i[1]+1, m)
+                    t_geo = pycolmap.estimate_calibrated_two_view_geometry(self.cam_obj, kps_2d[i[0]], self.cam_obj, kps_2d[i[1]], m).todict()
+                    self.db.add_two_view_geometry(
+                        image_id1 = i[0]+1, 
+                        image_id2 = i[1]+1, 
+                        matches = t_geo['inlier_matches'], 
+                        F = t_geo['F'], 
+                        E = t_geo['E'],
+                        H = t_geo['H'],
+                        config = t_geo['config'])
+
+        self.db.commit()
+
+    def reconstruction_sparse(self, dest, obj_name='object'):
+        print("\n\nSPARSE RECONSTUCTION....")
+        try:
+            os.mkdir(dest)
+        except:
+            shutil.rmtree(dest)
+            os.mkdir(dest)
+        self.__add_data()
+        maps = pycolmap.incremental_mapping(self.db_path, self.src, dest)
+        for key in maps:
+            maps[key].export_PLY(f'{dest}/{obj_name}_{key}.ply')
+        
+        if len(maps) > 1:
+            print("MERGING PLY FILES")
+            files = []
+            for file_name in os.listdir(dest):
+                if len(file_name) < 4 or os.path.splitext(file_name)[-1].lower() != '.ply':
+                    continue
+
+                file = plyfile.PlyData.read(os.path.join(dest, file_name))
+                for element in file.elements:
+                    files.append(element.data)
+            
+            merged_file = np.concatenate(files, -1)
+            merged_el = plyfile.PlyElement.describe(merged_file, "vertex")
+            plyfile.PlyData([merged_el]).write(f'{dest}/{obj_name}_merged.ply')
+        
+    def reconstruction_dense(self, dest, obj_name='object', model_folder='/0'):
+            print("\n\nDENSE RECONSTRUCTION")
+            mvs = f'{dest}/mvs'
+            dest += model_folder
+            try:
+                os.mkdir(mvs)
+            except:
+                shutil.rmtree(mvs)
+                os.mkdir(mvs)
+            try:
+                pycolmap.undistort_images(mvs, dest, self.src)
+                pycolmap.patch_match_stereo(mvs)
+                pycolmap.stereo_fusion(f'{dest}/{obj_name}_dense.ply', mvs)
+            except AttributeError as e:
+                print("PYCOLMAP NEEDS A CUDA BUILD")
```

Comment:
"PYCOLMAP NEEDS A CUDA BUILD, PLEASE USE THE COLMAP GUI FOR DENSE RECONSTRUCTION."